### PR TITLE
New API for sending ffi buffer in tcpsock without casting to string

### PR DESCRIPTION
**Pull request by dqminh**
This expose a way to set the tcp upstream socket's request buffer
manually, so we can call write later. The reason this is splitted into 2
functions is because i couldn't find a way to send cdata over in a
cfunction sanely.

Here is roughly how you can use them

```
local ffi = require "ffi"
local base = require "resty.core.base"

local C = ffi.C
local error = error
local errmsg = base.get_errmsg_ptr()
local FFI_OK = base.FFI_OK

ffi.cdef[[
int ngx_http_lua_socket_tcp_ffi_set_buf(ngx_http_request_t *r, void *u, void *cdata_ctx, size_t len, char **err);
]]

local function check_tcp(tcp)
  if not tcp or type(tcp) ~= "table" then
    return error("bad \"tcp\" argument")
  end
  tcp = tcp[1]
  if type(tcp) ~= "userdata" then
    return error("bad \"tcp\" argument")
  end
  return tcp
end

local function ffi_set_buffer(tcp, buffer, len)
  tcp = check_tcp(tcp)
  local r = base.get_request()
  if not r then
    return error("no request found")
  end
  local rc = C.ngx_http_lua_socket_tcp_ffi_set_buf(r, tcp, buffer, len, errmsg)
  if rc ~= FFI_OK then
    return nil, ffi.string(errmsg[0])
  end
  return true
end

local sock = ngx.socket.tcp()
sock:connect("unix:/tmp/go.sock")
sock:settimeout(1000)
ffi_set_buffer(sock, data, len)
sock:ffi_send()
```

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_socket_tcp.c [hunk 1]
```diff
@@ -27,6 +27,11 @@ static int ngx_http_lua_socket_tcp_sslhandshake(lua_State *L);
 static int ngx_http_lua_socket_tcp_receive(lua_State *L);
 static int ngx_http_lua_socket_tcp_receiveany(lua_State *L);
 static int ngx_http_lua_socket_tcp_send(lua_State *L);
+
+// send the exising buffer out non-blocking way.
+// buffer should be set using ngx_http_lua_socket_tcp_ffi_set_buf
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: Do not use C++-style comments in this module. Please see the C coding style guide for more details:

https://openresty.org/en/c-coding-style-guide.html

File: src/ngx_http_lua_socket_tcp.c [hunk 2]
```diff
@@ -2963,6 +2981,225 @@ ngx_http_lua_socket_tcp_send(lua_State *L)
     return lua_yield(L, 0);
 }
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: The nginx coding style requires the use of 2 blank lines to separate function definitions. See the guide mentioned above.

File: src/ngx_http_lua_socket_tcp.c [hunk 3]
```diff
@@ -2963,6 +2981,225 @@ ngx_http_lua_socket_tcp_send(lua_State *L)
...
+        return NGX_ERROR;
+    }
+
+    if (u->write_waiting) {
+        *err = "socket busy writing";
+        return NGX_ERROR;
+    }
+
+    if (u->raw_downstream
+        && (r->connection->buffered & NGX_HTTP_LOWLEVEL_BUFFERED)) {
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: When the if condition is in multiple lines, the `{` should be on a separate line by itself. See the guide mentioned above.

File: src/ngx_http_lua_socket_tcp.c [hunk 4]
```diff
@@ -2963,6 +2981,225 @@ ngx_http_lua_socket_tcp_send(lua_State *L)
...
+        return NGX_ERROR;
+    }
+
+    b = cl->buf;
+    b->last = ngx_copy(b->last, cdata_ctx, len);
+    u->request_bufs = cl;
+    u->request_len = len;
+    return NGX_OK;
+}
+
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: Ditto.

File: src/ngx_http_lua_socket_tcp.c [hunk 5]
```diff
@@ -295,6 +299,13 @@ ngx_http_lua_inject_socket_tcp_api(ngx_log_t *log, lua_State *L)
     lua_pushcfunction(L, ngx_http_lua_socket_tcp_settimeouts);
     lua_setfield(L, -2, "settimeouts"); /* ngx socket mt */
 
+#ifndef NGX_LUA_NO_FFI_API
+
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: We'd better remove this blank line.

File: src/ngx_http_lua_socket_tcp.c [hunk 6]
```diff
@@ -295,6 +299,13 @@ ngx_http_lua_inject_socket_tcp_api(ngx_log_t *log, lua_State *L)
...
+    lua_pushcfunction(L, ngx_http_lua_socket_tcp_ffi_send);
+    lua_setfield(L, -2, "ffi_send");
+
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: Ditto. Please address other similar places.

File: src/ngx_http_lua_socket_tcp.c [hunk 7]
```diff
@@ -295,6 +299,13 @@ ngx_http_lua_inject_socket_tcp_api(ngx_log_t *log, lua_State *L)
     lua_pushcfunction(L, ngx_http_lua_socket_tcp_settimeouts);
     lua_setfield(L, -2, "settimeouts"); /* ngx socket mt */
 
+#ifndef NGX_LUA_NO_FFI_API
+
+    lua_pushcfunction(L, ngx_http_lua_socket_tcp_ffi_send);
```
Comment (agentzh) [inline src/ngx_http_lua_socket_tcp.c]: As I mentioned, we should not implement this as a C function.

**Comment by zhuizhuhaomeng** [comment]
Rewrite this PR in https://github.com/openresty/lua-nginx-module/pull/2294.

**Pull request closed without merging**